### PR TITLE
[daint] Add Julia 1.5.3 and improve Julia admin path handling

### PR DESCRIPTION
**Pull request by omlins**
## The problem
Building and testing of certain packages from within the Julia Pkg manager can fail due to the Julia startup file. Either because 1) it gives an error if the DEPOT_PATH/LOAD_PATH were modified and the length does not correspond to length(USER_DEPOT_PATH) + length(ADMIN_DEPOT_PATH) or length(USER_LOAD_PATH) + length(ADMIN_LOAD_PATH)  or 2) the startup file is deactivated (according to the documentation, building happens without startup file by default except if julia is called explicitly with --startup-file=yes).

## Solution
1. The modules leave the DEPOT_PATH and LOAD_PATH essentially as defined by default by the Julia installation (introducing though mc and daint as well as using #.#.#).
2. The modules define the ADMIN_DEPOT_PATH and ADMIN_LOAD_PATH environment variables as before, with the exception that the standard Julia  installation depot paths  (/apps/daint/UES/omlins/julia-1.5.3/software/Julia/1.5.3-CrayGNU-20.08/local/share/julia and /apps/daint/UES/omlins/julia-1.5.3/software/Julia/1.5.3-CrayGNU-20.08/share/julia) and load path (@stdlib) are not added (as already in the DEPOT_PATH  and LOAD_PATH)
3. USER_DEPOT_PATH and USER_LOAD_PATH are not defined or just not used (and kept for possible other usage).
4. STD_DEPOT_PATH and STD_LOAD_PATH environment variables are additionally defined.
5. The startup file injects the entries of ADMIN_DEPOT_PATH  and ADMIN_LOAD_PATH to the DEPOT_PATH and LOAD_PATH after any user path, but before the standard Julia installation depot and load paths.
6. The startup file does nothing in certain conditions to ensure proper and consisting working in any case (Pkg build, test ...):  if paths empty (or only "@" for LOAD_PATH) or all entries in std path.

## Result
1. Standard usage of Julia is as before. 
2. For specific operations which are run with --startup-file=no  (implicitly or explicitly) the admin packages are simply deactivated. 
3. For specific operations which are run with --startup-file=yes  (implicitly or explicitly) the admin packages are also deactivated if the  conditions listed in point 6 above apply. Else, the admin paths are activated, but there is no constraint on the DEPOT_PATH/LOAD_PATH defined by the user or Pkg manager.
4. In addition, each user has the choice to easily completely deactivate the usage of admin packages  by creating a julia alias with --startup-file=no.

### CODE REVIEW

**Comment by omlins** [comment]
retest this please

**Comment by omlins** [comment]
The startup file does correctly inject the admin paths at the right place into `LOAD_PATH` and `DEPOT_PATH` and preserves the correct entry order (when it is deactivated, no injection happens and no error occurs):
```julia
omlins@nid02048:~> module load JuliaExtensions/1.5.3-CrayGNU-20.08
omlins@nid02048:~> julia --startup-file=no
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.5.3 (2020-11-09)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> LOAD_PATH
3-element Array{String,1}:
 ^[[A"@"
 "@#.#.#-daint-mc"
 "@stdlib"

julia> DEPOT_PATH
3-element Array{String,1}:
 "/users/omlins/.julia/1.5.3/daint-mc"
 "/apps/daint/UES/omlins/julia-1.5.3/software/Julia/1.5.3-CrayGNU-20.08/local/share/julia"
 "/apps/daint/UES/omlins/julia-1.5.3/software/Julia/1.5.3-CrayGNU-20.08/share/julia"

julia> 
omlins@nid02048:~> julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.5.3 (2020-11-09)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> LOAD_PATH
5-element Array{String,1}:
 "@"
 "@#.#.#-daint-mc"
 "/apps/daint/UES/omlins/julia-1.5.3/software/JuliaExtensions/1.5.3-CrayGNU-20.08/extensions/environments/1.5.3-daint-mc"
 "/apps/daint/UES/omlins/julia-1.5.3/software/Julia/1.5.3-CrayGNU-20.08/extensions/environments/1.5.3-daint-mc"
 "@stdlib"

julia> DEPOT_PATH
5-element Array{String,1}:
 "/users/omlins/.julia/1.5.3/daint-mc"
 "/apps/daint/UES/omlins/julia-1.5.3/software/JuliaExtensions/1.5.3-CrayGNU-20.08/extensions"
 "/apps/daint/UES/omlins/julia-1.5.3/software/Julia/1.5.3-CrayGNU-20.08/extensions"
 "/apps/daint/UES/omlins/julia-1.5.3/software/Julia/1.5.3-CrayGNU-20.08/local/share/julia"
 "/apps/daint/UES/omlins/julia-1.5.3/software/Julia/1.5.3-CrayGNU-20.08/share/julia"

julia> 
```

**Comment by omlins** [comment]
Here is the corresponding behaviour on Daint-gpu:
```julia-repl
omlins@nid02048:~> module load JuliaExtensions/1.5.3-CrayGNU-20.08-cuda 
omlins@nid02048:~> julia --startup-file=no
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.5.3 (2020-11-09)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> DEPOT_PATH
3-element Array{String,1}:
 "/users/omlins/.julia/1.5.3/daint-gpu"
 "/apps/daint/UES/omlins/julia-1.5.3/software/Julia/1.5.3-CrayGNU-20.08-cuda/local/share/julia"
 "/apps/daint/UES/omlins/julia-1.5.3/software/Julia/1.5.3-CrayGNU-20.08-cuda/share/julia"

julia> LOAD_PATH
3-element Array{String,1}:
 "@"
 "@#.#.#-daint-gpu"
 "@stdlib"

julia> 
omlins@nid02048:~> julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.5.3 (2020-11-09)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> DEPOT_PATH
5-element Array{String,1}:
 "/users/omlins/.julia/1.5.3/daint-gpu"
 "/apps/daint/UES/omlins/julia-1.5.3/software/JuliaExtensions/1.5.3-CrayGNU-20.08-cuda/extensions"
 "/apps/daint/UES/omlins/julia-1.5.3/software/Julia/1.5.3-CrayGNU-20.08-cuda/extensions"
 "/apps/daint/UES/omlins/julia-1.5.3/software/Julia/1.5.3-CrayGNU-20.08-cuda/local/share/julia"
 "/apps/daint/UES/omlins/julia-1.5.3/software/Julia/1.5.3-CrayGNU-20.08-cuda/share/julia"

julia> LOAD_PATH
5-element Array{String,1}:
 "@"
 "@#.#.#-daint-gpu"
 "/apps/daint/UES/omlins/julia-1.5.3/software/JuliaExtensions/1.5.3-CrayGNU-20.08-cuda/extensions/environments/1.5.3-daint-gpu"
 "/apps/daint/UES/omlins/julia-1.5.3/software/Julia/1.5.3-CrayGNU-20.08-cuda/extensions/environments/1.5.3-daint-gpu"
 "@stdlib"

julia> 
```

**Comment by omlins** [comment]
FFTW package successfully added, built and tested from within the Pkg manager (previously causing issues):
```julia
(@1.5.3-daint-gpu) pkg> add FFTW
  Resolving package versions...
  Installed FFTW_jll ──────── v3.3.9+7
  Installed IntelOpenMP_jll ─ v2018.0.3+2
  Installed MKL_jll ───────── v2020.2.254+0
  Installed AbstractFFTs ──── v1.0.0
  Installed FFTW ──────────── v1.3.1
Downloading artifact: IntelOpenMP
Downloading artifact: FFTW
Updating `~/.julia/1.5.3/daint-gpu/environments/1.5.3-daint-gpu/Project.toml`
  [7a1cc6ca] + FFTW v1.3.1
Updating `~/.julia/1.5.3/daint-gpu/environments/1.5.3-daint-gpu/Manifest.toml`
  [621f4979] + AbstractFFTs v1.0.0
  [56f22d72] + Artifacts v1.3.0
  [7a1cc6ca] + FFTW v1.3.1
  [f5851436] + FFTW_jll v3.3.9+7
  [1d5cc7b8] + IntelOpenMP_jll v2018.0.3+2
  [692b3bcd] + JLLWrappers v1.2.0
  [856f044c] + MKL_jll v2020.2.254+0
  [189a3867] + Reexport v1.0.0
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [b77e0a4c] + InteractiveUtils
  [76f85450] + LibGit2
  [8f399da3] + Libdl
  [37e2e46d] + LinearAlgebra
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [44cfe95a] + Pkg
  [de0858da] + Printf
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [ea8e919c] + SHA
  [9e88b42a] + Serialization
  [6462fe0b] + Sockets
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode
   Building FFTW → `~/.julia/1.5.3/daint-gpu/packages/FFTW/eADNB/deps/build.log`

(@1.5.3-daint-gpu) pkg> test FFTW
    Testing FFTW
Status `/tmp/jl_Fc34UK/Project.toml`
  [621f4979] AbstractFFTs v1.0.0
  [7a1cc6ca] FFTW v1.3.1
  [f5851436] FFTW_jll v3.3.9+7
  [1d5cc7b8] IntelOpenMP_jll v2018.0.3+2
  [856f044c] MKL_jll v2020.2.254+0
  [189a3867] Reexport v1.0.0
  [8f399da3] Libdl
  [37e2e46d] LinearAlgebra
  [8dfed614] Test
Status `/tmp/jl_Fc34UK/Manifest.toml`
  [621f4979] AbstractFFTs v1.0.0
  [56f22d72] Artifacts v1.3.0
  [7a1cc6ca] FFTW v1.3.1
  [f5851436] FFTW_jll v3.3.9+7
  [1d5cc7b8] IntelOpenMP_jll v2018.0.3+2
  [692b3bcd] JLLWrappers v1.2.0
  [856f044c] MKL_jll v2020.2.254+0
  [189a3867] Reexport v1.0.0
  [2a0f44e3] Base64
  [ade2ca70] Dates
  [8ba89e20] Distributed
  [b77e0a4c] InteractiveUtils
  [76f85450] LibGit2
  [8f399da3] Libdl
  [37e2e46d] LinearAlgebra
  [56ddb016] Logging
  [d6f4376e] Markdown
  [44cfe95a] Pkg
  [de0858da] Printf
  [3fa0cd96] REPL
  [9a3f8284] Random
  [ea8e919c] SHA
  [9e88b42a] Serialization
  [6462fe0b] Sockets
  [8dfed614] Test
  [cf7118a7] UUIDs
  [4ec0a83e] Unicode
Test Summary:   | Pass  Total
colmajorstrides |    1      1
Test Summary: | Pass  Total
fft           | 1333   1333
Test Summary: | Pass  Total
rfft/rfftn    |  387    387
Test Summary:               |
Self-test: Complex{Float32} | No tests
Test Summary:               |
Self-test: Complex{Float64} | No tests
Test Summary:                                                    | Pass  Total
Inversion, scaling, and pre-allocated variants: Complex{Float32} |   40     40
Test Summary:                                                    | Pass  Total
Inversion, scaling, and pre-allocated variants: Complex{Float64} |   40     40
Test Summary:                           |
Base Julia issue #9772, with size (10,) | No tests
Test Summary:                              | Broken  Total
Base Julia issue #9772, with size (10, 12) |     15     15
Test Summary:           | Pass  Total
Base Julia issue #17896 |    4      4
Test Summary:                         | Pass  Total
Discrete cosine transform (DCT) tests |  337    337
Test Summary:  | Pass  Total
UNALIGNED flag |   74     74
Test Summary:           | Pass  Total
Base Julia issue #19892 |    3      3
    Testing FFTW tests passed 

(@1.5.3-daint-gpu) pkg> 
```

**Review by twrobinson — CHANGES_REQUESTED** [review]
Please add an updated JupyterLab 1.2.16 to the PR, as the FFT test by itself works at the command line currently.

**Comment by omlins** [comment]
`jupyterlab-1.2.16-CrayGNU-20.08-batchspawner.eb` (unmodified - as on master) fails to build on Daint with the following error:
```
== 2021-02-02 13:07:22,539 pythonpackage.py:185 INFO Determining pip version...
== 2021-02-02 13:07:22,539 run.py:222 INFO running cmd: pip --version
== 2021-02-02 13:07:23,262 pythonpackage.py:193 INFO Found pip version: 20.2.2
== 2021-02-02 13:07:23,262 pythonpackage.py:406 INFO Found pip version 20.2.2, OK
== 2021-02-02 13:07:23,263 run.py:222 INFO running cmd:  pip install --prefix=/apps/daint/UES/omlins/julia-1.5.3/software/jupyterlab/1.2.16-CrayGNU-20.08-batchspawner  --no-deps  --ignore-installed  --no-build-isolation  .
== 2021-02-02 13:07:25,189 build_log.py:169 ERROR EasyBuild crashed with an error (at easybuild/software/EasyBuild/4.3.2/lib/python2.7/site-packages/easybuild/base/exceptions.py:124 in __init__): cmd " pip install --prefix=/apps/daint/UES/omlins/julia-1.5.3/software/jupyterlab/1.2.16-CrayGNU-20.08-batchspawner  --no-deps  --ignore-installed  --no-build-isolation  ." exited with exit code 1 and output:
Processing /apps/daint/UES/omlins/julia-1.5.3/jupyterlab/1.2.16/CrayGNU-20.08-batchspawner/Pillow/Pillow-7.0.0
ERROR: Could not install packages due to an EnvironmentError: [('/apps/daint/UES/omlins/julia-1.5.3/jupyterlab/1.2.16/CrayGNU-20.08-batchspawner/Pillow/Pillow-7.0.0/Tests/images/truncated_image.png', '/run/user/24345/easybuild/tmp/eb-Oyr_AH/pip-req-build-kpuaffay/Tests/images/truncated_image.png', "[Errno 11] Resource temporarily unavailable: '/apps/daint/UES/omlins/julia-1.5.3/jupyterlab/1.2.16/CrayGNU-20.08-batchspawner/Pillow/Pillow-7.0.0/Tests/images/truncated_image.png' -> '/run/user/24345/easybuild/tmp/eb-Oyr_AH/pip-req-build-kpuaffay/Tests/images/truncated_image.png'"), ('/apps/daint/UES/omlins/julia-1.5.3/jupyterlab/1.2.16/CrayGNU-20.08-batchspawner/Pillow/Pillow-7.0.0/Tests/images/sgi_overrun.bin', '/run/user/24345/easybuild/tmp/eb-Oyr_AH/pip-req-build-kpuaffay/Tests/images/sgi_overrun.bin', "[Errno 11] Resource temporarily unavailable: '/apps/daint/UES/omlins/julia-1.5.3/jupyterlab/1.2.16/CrayGNU-20.08-batchspawner/Pillow/Pillow-7.0.0/Tests/images/sgi_overrun.bin' -> '/run/user/24345/easybuild/tmp/eb-Oyr_AH/pip-req-build-kpuaffay/Tests/images/sgi_overrun.bin'")]
```
@twrobinson  can you have a look?

**Comment by twrobinson** [comment]
> `jupyterlab-1.2.16-CrayGNU-20.08-batchspawner.eb` (unmodified - as on master) fails to build on Daint with the following error:
> 
> ```
> == installing extension Pillow 7.0.0 (66/91)...
> == FAILED: Installation ended unsuccessfully (build directory: /apps/daint/UES/omlins/julia-1.5.3/jupyterlab/1.2.16/CrayGNU-20.08-batchspawner): build failed (first 300 chars): cmd " pip install --prefix=/apps/daint/UES/omlins/julia-1.5.3/software/jupyterlab/1.2.16-CrayGNU-20.08-batchspawner  --no-deps  --ignore-installed  --no-build-isolation  ." exited with exit code 1 and output:
> Processing /apps/daint/UES/omlins/julia-1.5.3/jupyterlab/1.2.16/CrayGNU-20.08-batchspawner/ (took 13 min 56 sec)
> == Results of the build can be found in the log file(s) /run/user/24345/easybuild/tmp/eb-Oyr_AH/easybuild-jupyterlab-1.2.16-20210202.125329.uLOTk.log
> ERROR: Build of /apps/daint/UES/omlins/julia-1.5.3/production/easybuild/easyconfigs/j/jupyterlab/jupyterlab-1.2.16-CrayGNU-20.08-batchspawner.eb failed (err: 'build failed (first 300 chars): cmd " pip install --prefix=/apps/daint/UES/omlins/julia-1.5.3/software/jupyterlab/1.2.16-CrayGNU-20.08-batchspawner  --no-deps  --ignore-installed  --no-build-isolation  ." exited with exit code 1 and output:\nProcessing /apps/daint/UES/omlins/julia-1.5.3/jupyterlab/1.2.16/CrayGNU-20.08-batchspawner/')
> ```
> 
> Here is the probably relevant part from the end of the build log:
> 
> ```
> == 2021-02-02 13:07:22,539 pythonpackage.py:185 INFO Determining pip version...
> == 2021-02-02 13:07:22,539 run.py:222 INFO running cmd: pip --version
> == 2021-02-02 13:07:23,262 pythonpackage.py:193 INFO Found pip version: 20.2.2
> == 2021-02-02 13:07:23,262 pythonpackage.py:406 INFO Found pip version 20.2.2, OK
> == 2021-02-02 13:07:23,263 run.py:222 INFO running cmd:  pip install --prefix=/apps/daint/UES/omlins/julia-1.5.3/software/jupyterlab/1.2.16-CrayGNU-20.08-batchspawner  --no-deps  --ignore-installed  --no-build-isolation  .
> == 2021-02-02 13:07:25,189 build_log.py:169 ERROR EasyBuild crashed with an error (at easybuild/software/EasyBuild/4.3.2/lib/python2.7/site-packages/easybuild/base/exceptions.py:124 in __init__): cmd " pip install --prefix=/apps/daint/UES/omlins/julia-1.5.3/software/jupyterlab/1.2.16-CrayGNU-20.08-batchspawner  --no-deps  --ignore-installed  --no-build-isolation  ." exited with exit code 1 and output:
> Processing /apps/daint/UES/omlins/julia-1.5.3/jupyterlab/1.2.16/CrayGNU-20.08-batchspawner/Pillow/Pillow-7.0.0
> ERROR: Could not install packages due to an EnvironmentError: [('/apps/daint/UES/omlins/julia-1.5.3/jupyterlab/1.2.16/CrayGNU-20.08-batchspawner/Pillow/Pillow-7.0.0/Tests/images/truncated_image.png', '/run/user/24345/easybuild/tmp/eb-Oyr_AH/pip-req-build-kpuaffay/Tests/images/truncated_image.png', "[Errno 11] Resource temporarily unavailable: '/apps/daint/UES/omlins/julia-1.5.3/jupyterlab/1.2.16/CrayGNU-20.08-batchspawner/Pillow/Pillow-7.0.0/Tests/images/truncated_image.png' -> '/run/user/24345/easybuild/tmp/eb-Oyr_AH/pip-req-build-kpuaffay/Tests/images/truncated_image.png'"), ('/apps/daint/UES/omlins/julia-1.5.3/jupyterlab/1.2.16/CrayGNU-20.08-batchspawner/Pillow/Pillow-7.0.0/Tests/images/sgi_overrun.bin', '/run/user/24345/easybuild/tmp/eb-Oyr_AH/pip-req-build-kpuaffay/Tests/images/sgi_overrun.bin', "[Errno 11] Resource temporarily unavailable: '/apps/daint/UES/omlins/julia-1.5.3/jupyterlab/1.2.16/CrayGNU-20.08-batchspawner/Pillow/Pillow-7.0.0/Tests/images/sgi_overrun.bin' -> '/run/user/24345/easybuild/tmp/eb-Oyr_AH/pip-req-build-kpuaffay/Tests/images/sgi_overrun.bin'")]
> ```
> 
> @twrobinson can you have a look?

It built ok for me (https://github.com/eth-cscs/production/pull/2132).

**Comment by omlins** [comment]
retest this please

**Comment by omlins** [comment]
I added the jupyterlab recipe here now as it worked for @twrobinson  when letting Jenkins build it. Then, it should build this now too.

**Comment by omlins** [comment]
retest this please

**Comment by omlins** [comment]
Julia needs to be built on the elogin nodes as it sets the paths based on where it is built in order to be able to have separate installations for daint and dom. EB testing happens however on the compute nodes, leading to different paths.
Built successfully jupyterlab on Daint and tested on jupyter.cscs.ch. All usual tests succeeded as well as adding, building and testing FFTW from within the Pkg manager succeeded, resolving all reported issues.

**Review by twrobinson — CHANGES_REQUESTED** [review]
I suggest changing this PR to be for Dom, in which case the files of relevance are 20.11. Daint will move to 20.11 next week, so there is little point in targetting 20.08. Let's go directly with a PR targetting Dom and 20.11. You can add my changes from #2133 into this PR.

File: easybuild/easyconfigs/j/Julia/Julia-1.5.3-CrayGNU-20.08-cuda.eb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+dependencies = [
+    ('craype-accel-nvidia60', EXTERNAL_MODULE),
+]
+
+arch_name = 'gpu'
+exts_defaultclass = 'JuliaPackage'
+
+exts_list = [
+    ('MPI.jl', '0.16.1', {'mpiexec': 'srun', 'mpiexec_args': '-C gpu', 'source_tmpl': 'v0.16.1.tar.gz', 'source_urls': ['https://github.com/JuliaParallel/MPI.jl/archive/']}),
+    ('CUDA.jl', '2.4.1', {'source_tmpl': 'v2.4.1.tar.gz', 'source_urls': ['https://github.com/JuliaGPU/CUDA.jl/archive/']}),
```
Comment (victorusu) [inline easybuild/easyconfigs/j/Julia/Julia-1.5.3-CrayGNU-20.08-cuda.eb:27]: I wonder if we should add the FFTW package here.

**Review by victorusu — APPROVED** [review]
`lgtm`. I agree with @twrobinson that we should focus on 20.11, but it doesn't hurt to have 20.08 also corrected.
I added a comment about extending the dependency list to add FFTW which - if my mind doesn't fail me - is the culprit for all these changes.

File: easybuild/easyconfigs/j/Julia/Julia-1.5.3-CrayGNU-20.08-cuda.eb (hunk 1 above)
Comment (omlins) [reply easybuild/easyconfigs/j/Julia/Julia-1.5.3-CrayGNU-20.08-cuda.eb:27]: Agreed offline that for now we do not want to add FFTW in here.

**Comment by omlins** [comment]
I am preparing a PR for 20.11. However, this PR should get merged now as Julia is working as it should with 20.08 and in order to keep independent steps separate as much as possible.

**Comment by victorusu** [comment]
@omlins, for me it is fine. let's wait for @twrobinson...

**Review by twrobinson — APPROVED** [review]

**Comment by twrobinson** [comment]
Fine to merge

**Pull request merged**